traits/observers/_list_item_observer.py

```diff
@@ -0,0 +1,221 @@
+                return
+            raise ValueError(
+                "Expected a TraitList to be observed, "
+                "got {!r} (type: {!r})".format(object, type(object)))
+
+        yield from object
+
+    def get_notifier(self, handler, target, dispatcher):
+        """ Return a notifier for calling the user handler with the change
+        event.
+
+        Returns
+        -------
+        notifier : TraitEventNotifier
+        """
+        # Unlike CTrait, when default list is created, there isn't a change
+        # event where the old value is Uninitialized.
+        return TraitEventNotifier(
+            handler=handler,
+            target=target,
+            dispatcher=dispatcher,
+            event_factory=list_event_factory,
+            prevent_event=lambda event: False,
+        )
+
+    def get_maintainer(self, graph, handler, target, dispatcher):
+        """ Return a notifier for maintaining downstream observers when
+        a list is mutated.
+
+        Parameters
+        ----------
+        graph : ObserverGraph
+            Description for the *downstream* observers, i.e. excluding self.
+        handler : callable
+            User handler.
+        target : object
+            Object seen by the user as the owner of the observer.
+        dispatcher : callable
+            Callable for dispatching the handler.
+
+        Returns
+        -------
+        notifier : ObserverChangeNotifier
+        """
+        return ObserverChangeNotifier(
+            observer_handler=_observer_change_handler,
+            event_factory=list_event_factory,
+            prevent_event=lambda event: False,
+            graph=graph,
+            handler=handler,
+            target=target,
+            dispatcher=dispatcher,
+        )
+
+    def iter_extra_graphs(self, graph):
+        """ Yield new ObserverGraph to be contributed by this observer.
+
+        Parameters
+        ----------
+        graph : ObserverGraph
+            The graph this observer is part of.
+
+        Yields
+        ------
+        ObserverGraph
+        """
+        # Unlike CTrait, no need to handle trait_added
+        yield from ()
+
+
+def _observer_change_handler(event, graph, handler, target, dispatcher):
+    """ Handler for maintaining observers. Used by ObserverChangeNotifier.
+
+    The downstream notifiers are removed from items removed from the list.
+    Likewise, downstream notifiers are added to items added to the list.
+
+    Parameters
+    ----------
+    event : ListChangeEvent
+        Change event that triggers the maintainer.
+    graph : ObserverGraph
+        Description for the *downstream* observers, i.e. excluding self.
+    handler : callable
+        User handler.
+    target : object
+        Object seen by the user as the owner of the observer.
+    dispatcher : callable
+        Callable for dispatching the handler.
+    """
+    for removed_item in event.removed:
+        add_or_remove_notifiers(
```

Comment:
The `add_or_remove_notifiers(..., remove=True)` feels awkward here; I think we probably want `add_notifiers` and `remove_notifiers` functions instead, for clarity and simplicity. (This is one of Guido's rules for Pythonic interfaces: if you have a boolean flag that's almost always passed as a constant in practice, you probably want two separate functions instead.)

Reply:
On the one hand, I agree with that. I prefer `observe` and `unobserve`.

However, that would go against local consistency with `on_trait_change`. This pattern is not just in traits, but also in traitsui and pyface.

On the other hand, this `add_or_remove_notifiers` is only used internally so it would also be fine to separate them.  On the other other hand, it is much easier to ensure the "add" operation and "remove" operation are the exact opposite if they are implemented in the same function. 

All in all, we might end up having an intermediate `add_notifiers` and `remove_notifiers` that wraps an `add_or_remove_notifiers`, and then the `add_notifiers` and `remove_notifiers` being indirectly called by `observe(remove: boolean)`. 🤯

Reply:
In fact, separate `add_notifiers` and `remove_notifiers` are what I initially had, but then I ended up having another if-else-branch in `observe` wrapping something that has another if-else-branch (the code that does the mutation on the notifier list), so I gave up on that. The pattern has to happen consistently for this to be helpful.

Reply:
Okay, let's leave as-is for now.